pkg/providers/cloudstack/apibuilder.go

```diff
@@ -0,0 +1,119 @@
+package cloudstack
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	cloudstackv1 "sigs.k8s.io/cluster-api-provider-cloudstack/api/v1beta2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/clusterapi"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	anywhereTypes "github.com/aws/eks-anywhere/pkg/types"
+)
+
+const (
+	CloudStackMachineTemplateKind      = "CloudStackMachineTemplate"
+	CloudStackInfrastructureAPIVersion = "infrastructure.cluster.x-k8s.io/v1beta2"
+)
+
+func machineDeployment(clusterSpec *cluster.Spec, workerNodeGroupConfig v1alpha1.WorkerNodeGroupConfiguration, kubeadmConfigTemplate *bootstrapv1.KubeadmConfigTemplate, cloudstackMachineTemplate *cloudstackv1.CloudStackMachineTemplate) clusterv1.MachineDeployment {
+	return clusterapi.MachineDeployment(clusterSpec, workerNodeGroupConfig, kubeadmConfigTemplate, cloudstackMachineTemplate)
+}
+
+func MachineDeployments(clusterSpec *cluster.Spec, kubeadmConfigTemplates map[string]*bootstrapv1.KubeadmConfigTemplate, machineTemplates map[string]*cloudstackv1.CloudStackMachineTemplate) map[string]*clusterv1.MachineDeployment {
+	m := make(map[string]*clusterv1.MachineDeployment, len(clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations))
+
+	for _, workerNodeGroupConfig := range clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		deployment := machineDeployment(clusterSpec, workerNodeGroupConfig,
+			kubeadmConfigTemplates[workerNodeGroupConfig.Name],
+			machineTemplates[workerNodeGroupConfig.Name],
+		)
+		m[workerNodeGroupConfig.Name] = &deployment
+	}
+	return m
+}
+
+func fillMachineTemplateAnnotations(machineConfig *v1alpha1.CloudStackMachineConfigSpec) map[string]string {
+	annotations := make(map[string]string, 0)
+	if machineConfig.DiskOffering != nil {
+		annotations[fmt.Sprintf("mountpath.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.MountPath
+		annotations[fmt.Sprintf("device.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Device
+		annotations[fmt.Sprintf("filesystem.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Filesystem
+		annotations[fmt.Sprintf("label.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Label
+	}
+
+	if machineConfig.Symlinks != nil {
+		links := make([]string, 0)
+		for key := range machineConfig.Symlinks {
+			links = append(links, fmt.Sprintf("%s:%s", key, machineConfig.Symlinks[key]))
+		}
+		// sorting for unit test determinism
+		sort.Strings(links)
```

Comment:
Instead of needing to sort in the business logic, can we just change our verification logic in our unit test to use a hash set?

Reply:
The unit tests in provider/cloudstack.go are strictly validating against yaml "expected" testdata at the moment which don't have this capability :(

Reply:
Can we create an action item to address this? I would really like to avoid doing something for a unit test in our business logic. Not sure if we do this today already, but ideally the symlinks passed into this method would already be sorted for that case. Are we able to control that?

Reply:
The symlinks are passed in as a standard unsorted hashmap and they get sorted arbitrarily when you iterate over them. The other option would be only testing for a single symlink but then we do not get validation on the multi-symlink formatting, which feels like missed coverage and risky. I agree about adding logic for unit tests to business logic is messy, but it seems harmless in any real scenario in my opinion so I'm inclined to keep it for now.